sdk/core/src/main/java/com/klaviyo/coresdk/networking/requests/KlaviyoApiRequest.kt

```diff
@@ -0,0 +1,230 @@
+     *
+     * @return
+     */
+    fun toJson(): String = JSONObject()
+        .accumulate(PATH_JSON_KEY, urlPath)
+        .accumulate(METHOD_JSON_KEY, method.name)
+        .accumulate(TIME_JSON_KEY, time)
+        .accumulate(UUID_JSON_KEY, uuid)
+        .accumulate(HEADERS_JSON_KEY, headers)
+        .accumulate(QUERY_JSON_KEY, query)
+        .accumulate(BODY_JSON_KEY, body)
+        .toString()
+
+    /**
+     * To facilitate deduplication, we will treat UUID as a unique identifier
+     * such that even if a request is deserialized from JSON it is still
+     * "equal" to its original object, regardless of instance or subclass
+     *
+     * @param other
+     * @return
+     */
+    override fun equals(other: Any?): Boolean = when (other) {
+        is KlaviyoApiRequest -> equals(this, other)
+        else -> super.equals(other)
+    }
+
+    companion object {
+        private const val PATH_JSON_KEY = "url_path"
+        private const val METHOD_JSON_KEY = "method"
+        private const val TIME_JSON_KEY = "time"
+        private const val UUID_JSON_KEY = "uuid"
+        private const val HEADERS_JSON_KEY = "headers"
+        private const val QUERY_JSON_KEY = "query"
+        private const val BODY_JSON_KEY = "body"
+
+        /**
+         * Construct a request from a JSON object
+         *
+         * Returns [KlaviyoApiRequest] no matter what subclass it was before encoding.
+         * This is functionally the same and saves us some trouble
+         *
+         * @param json
+         * @return
+         */
+        fun fromJson(json: JSONObject): KlaviyoApiRequest {
+            val urlPath = json.getString(PATH_JSON_KEY)
+            val method = when (json.getString(METHOD_JSON_KEY)) {
+                RequestMethod.POST.name -> RequestMethod.POST
+                else -> RequestMethod.GET
+            }
+            val time = json.getString(TIME_JSON_KEY)
+            val uuid = json.getString(UUID_JSON_KEY)
+
+            return KlaviyoApiRequest(urlPath, method, time, uuid).apply {
+                headers = json.getJSONObject(HEADERS_JSON_KEY).let {
+                    it.keys().asSequence().associateWith { k -> it.getString(k) }
+                }
+                query = json.getJSONObject(QUERY_JSON_KEY).let {
+                    it.keys().asSequence().associateWith { k -> it.getString(k) }
+                }
+                body = json.optJSONObject(BODY_JSON_KEY)
+            }
+        }
+
+        /**
+         * Compare two requests purely by their UUIDs
+         *
+         * @param request1
+         * @param request2
+         * @return
+         */
+        fun equals(request1: KlaviyoApiRequest, request2: KlaviyoApiRequest): Boolean =
+            request1.uuid == request2.uuid
+    }
+
+    /**
+     * Compiles the base url, path and query data into a [URL] object
+     */
+    val url: URL
+        get() {
+            val baseUrl = Klaviyo.Registry.config.baseUrl
```

Comment:
So with the original structure, one thought I had was that I was building an abstract `NetworkRequest` class separate from whatever concrete implementation decides that it's a Klaviyo request. So I had a few layers of inheritance but thought if there were use cases in the future, maybe we'd have separate request objects that handled authentication/constants/specific-payload-building for non-Klaviyo entities.

Reply:
I may have just over-prepped though. I don't think we have any reason to be making network requests to any other services through this SDK.

But thats the reasons around why it was so layered originally.

Reply:
I figured, but also figured I didn't see us communicating with other services so subtracted a bit of the complication that way, its pretty easy to add it back too.
The main reason that I took out the abstraction layer was that this structure made it easy to deserialize a JSON representation of any API request object and have it be _functionally_ equivalent even though I don't deserialize into the subclass.

Reply:
Yup thats all fine with me! Just thought I'd specify what was going through my head at the time!